File: src/AssignmentSolver.cpp
/*
 * Copyright 2020 Maximilian Azendorf
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "AssignmentSolver.h"

#ifndef USE_CBC
#define USE_CBC
#endif
#include <ortools/linear_solver/linear_solver.h>

#include "Constraints.h"

#include <utility>

set<pair<int, int>> AssignmentSolver::get_blocked_constraint_edges(shared_ptr<Scheduling const> const& scheduling)
{
    set<pair<int, int>> blockedEdges;

    for(Constraint constraint : _staticData->constraints)
    {
        switch(constraint.type())
        {
            case ChooserIsInChoice:
            {
                int s = scheduling->slot_of(constraint.right());
                if (s < 0) continue;

                int from = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_chooser(constraint.left(), s));

                for(int w = 0; w < _inputData->choice_count(); w++)
                {
                    if(constraint.right() == w || scheduling->slot_of(w) != s) continue;

                    int to = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_choice(w));
                    blockedEdges.insert(std::make_pair(from, to));
                }
                break;
            }
            case ChooserIsNotInChoice:
            {
                for(int s = 0; s < _inputData->slot_count(); s++)
                {
                    int from = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_chooser(constraint.left(), s));
                    int to = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_choice(constraint.right()));
                    blockedEdges.insert(std::make_pair(from, to));
                }
                break;
            }
            case ChoicesHaveSameChoosers: // This is handled in create_edge_groups with Constraints::get_dependent_choices.
            case ChoosersHaveSameChoices: // This is also handled in create_edge_groups.
            {
                break;
            }
            default:
            {
                throw std::logic_error("This kind of constraint is not compatible with the min cost flow solver.");
            }
        }
    }

    return blockedEdges;
}

void AssignmentSolver::create_edge_groups(const_ptr<Scheduling> const& scheduling, MipFlow<flowid, flowid>& flow)
{
    // Edge groups for ChoosersHaveSameChoices
    //
    for(Constraint constraint : _inputData->assignment_constraints())
    {
        if(constraint.type() != ChoosersHaveSameChoices) continue;

        for (int s = 0; s < _inputData->slot_count(); s++)
        {
            for (int w = 0; w < _inputData->choice_count(); w++)
            {
                int from1 = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_chooser(constraint.left(), s));
                int to1 = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_choice(w));
                int from2 = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_chooser(constraint.right(), s));
                int to2 = _staticData->baseFlow.nodes().at(MipFlowStaticData::node_choice(w));

                vector<flowid> edgeGroup { MipFlowStaticData::edge_id(from1, to1), MipFlowStaticData::edge_id(from2, to2) };
                flow.create_edge_group_or_block_edges(edgeGroup.begin(), edgeGroup.end());
            }
        }
    }

    // Edge groups for dependent choices
    //
    for(vector<int> const& group : Constraints::get_dependent_choices(
            _inputData->assignment_constraints(),
            _inputData->choice_count()))
    {
        if(group.size() == 1) continue;

        for(int p = 0; p < _inputData->chooser_count(); p++)
        {
            vector<flowid> edgeGroup;
            for(int w : group)
            {
                int s = scheduling->slot_of(w);
                if (s < 0) continue;

                int from = flow.nodes().at(MipFlowStaticData::node_chooser(p, s));
                int to = flow.nodes().at(MipFlowStaticData::node_choice(w));
                edgeGroup.push_back(MipFlowStaticData::edge_id(from, to));
            }

            flow.create_edge_group_or_block_edges(edgeGroup.begin(), edgeGroup.end());
        }
    }
}

const_ptr<Assignment> AssignmentSolver::solve_with_limit(const_ptr<Scheduling> const& scheduling,
                                                                int preferenceLimit,
                                                                op::MPSolver& solver)
{
    MipFlow<flowid, flowid> flow(_staticData->baseFlow);

    for(int p = 0; p < _inputData->chooser_count(); p++)
    {
        for(int s = 0; s < _inputData->slot_count(); s++)
        {
            flow.set_supply(flow.nodes().at(MipFlowStaticData::node_chooser(p, s)), 1);
        }
    }

    for(int w = 0; w < _inputData->choice_count(); w++)
    {
        if (scheduling->slot_of(w) == Scheduling::NOT_SCHEDULED) continue;
        flow.set_supply(flow.nodes().at(MipFlowStaticData::node_choice(w)), -_inputData->choice(w).min);
    }

    for(int s = 0; s < _inputData->slot_count(); s++)
    {
        // Count the number of choosers that will already be absorbed by the choice nodes.
        //
        int coveredChoosers = 0;
        for(int w = 0; w < _inputData->choice_count(); w++)
        {
            if(scheduling->slot_of(w) != s) continue;
            coveredChoosers += _inputData->choice(w).min;
        }

        flow.set_supply(
                flow.nodes().at(MipFlowStaticData::node_slot(s)),
                -(_inputData->chooser_count() - coveredChoosers));
    }

    vector<int> edgesCap;
    vector<long> edgesCost;
    map<pair<int, int>, int> edgesIdx;
    int nextEdgeIdx = 0;

    for(int p = 0; p < _inputData->chooser_count(); p++)
    {
        for(int s = 0; s < _inputData->slot_count(); s++)
        {
            for(int w = 0; w < _inputData->choice_count(); w++)
            {
                if(scheduling->slot_of(w) != s || _inputData->chooser(p).preferences[w] > preferenceLimit)
                    continue;

                edgesCap.push_back(1);
                edgesCost.push_back(
                        (long)pow(_inputData->chooser(p).preferences[w] + 1.0, _options->preference_exponent()));

                edgesIdx[std::make_pair(
                        flow.nodes().at(MipFlowStaticData::node_chooser(p, s)),
                        flow.nodes().at(MipFlowStaticData::node_choice(w)))]
                        = nextEdgeIdx++;
            }
        }
    }

    for(int w = 0; w < _inputData->choice_count(); w++)
    {
        for(int s = 0; s < _inputData->slot_count(); s++)
        {
            if(scheduling->slot_of(w) != s) continue;

            edgesCap.push_back(_inputData->choice(w).max - _inputData->choice(w).min);
            edgesCost.push_back(0);

            edgesIdx[std::make_pair(
                    flow.nodes().at(MipFlowStaticData::node_choice(w)),
                    flow.nodes().at(MipFlowStaticData::node_slot(s)))]
                    = nextEdgeIdx++;
        }
    }

    // Remove all blocked edges
    //
    auto blockedEdges = get_blocked_constraint_edges(scheduling);
    blockedEdges.insert(_staticData->blockedEdges.begin(), _staticData->blockedEdges.end());

    for(auto blocked : blockedEdges)
    {
        edgesIdx.erase(blocked);
    }

    for(auto edge : edgesIdx)
    {
        int start = edge.first.first;
        int end = edge.first.second;
        int cap = edgesCap[edge.second];
        long cost = edgesCost[edge.second];

        flow.add_edge(MipFlowStaticData::edge_id(start, end), start, end, cap, cost);
    }

    // Create edge groups
    //
    create_edge_groups(scheduling, flow);

    // ... and solve this instance
    //
    _lpCount++;
    if(!flow.solve(solver))
    {
        return nullptr;
    }

    // Now we have to extract the assignment solution from the min cost flow solution
    //
    vector<vector<int>> data(_inputData->chooser_count(), vector<int>(_inputData->slot_count(), -1));
    for(int p = 0; p < _inputData->chooser_count(); p++)
    {
        for(int s = 0; s < _inputData->slot_count(); s++)
        {
            for(int w = 0; w < _inputData->choice_count(); w++)
            {
                int from = flow.nodes().at(MipFlowStaticData::node_chooser(p, s));
                int to = flow.nodes().at(MipFlowStaticData::node_choice(w));

                if(flow.solution_value_at(MipFlowStaticData::edge_id(from, to)) == 1)
                {
                    data[p][s] = w;
                }
            }
        }
    }

    return std::make_shared<Assignment const>(_inputData, data);
}

const_ptr<Assignment> AssignmentSolver::solve(const_ptr<Scheduling const> const& scheduling)
{
    auto solver = op::MPSolver("solver", op::MPSolver::CBC_MIXED_INTEGER_PROGRAMMING);

    int prefIdx = 0;
    for(; prefIdx < _inputData->preference_levels().size(); prefIdx++)
    {
        if(_inputData->preference_levels().at(prefIdx) == _csAnalysis->preference_bound()) break;
    }

    int minIdx = prefIdx;
    int maxIdx = _inputData->preference_levels().size();

    shared_ptr<Assignment const> bestAssignment = nullptr;

    if(!_options->greedy())
    {
        // We do binary search through all possible preference limits to find the lowest one.
        //
        do
        {
            int prefLimit = _inputData->preference_levels().at(prefIdx);
            auto assignment = solve_with_limit(scheduling, prefLimit, solver);
            if (assignment == nullptr)
            {
                minIdx = prefIdx + 1;
            }
            else
            {
                bestAssignment = assignment;
                maxIdx = prefIdx - 1;
            }
            prefIdx = (maxIdx + minIdx) / 2;
        } while (maxIdx > minIdx);
    }
    else
    {
        // In greedy mode, we don't set a preference limit; just solve it.
        //
        bestAssignment = solve_with_limit(scheduling, _inputData->max_preference(), solver);
    }

    return bestAssignment;
}

AssignmentSolver::AssignmentSolver(const_ptr<InputData> inputData,
                                   const_ptr<CriticalSetAnalysis> csAnalysis,
                                   const_ptr<MipFlowStaticData> staticData,
                                   const_ptr<Options> options,
                                   cancel_token cancellation)
    : _inputData(std::move(inputData)),
    _csAnalysis(std::move(csAnalysis)),
    _staticData(std::move(staticData)),
    _options(std::move(options)),
    _cancellation(std::move(cancellation))
{
}

int AssignmentSolver::lp_count() const
{
    return _lpCount;
}

File: src/assignment_solver.rs
use std::collections::HashSet;
use std::time::{Duration, SystemTime};

use crate::shotgun_solver::ProgressReporter;
use crate::{
    Assignment, ConstraintType, MipFlow, MipFlowStaticData, Options, PreparedProblem, Scheduling,
};
use crate::status;

/// Computes an optimal assignment for a fixed scheduling.
#[derive(Debug)]
pub struct AssignmentSolver<'a> {
    problem: &'a PreparedProblem,
    options: &'a Options,
    progress_reporter: ProgressReporter,
    pub(crate) lp_count: usize,
    cached_schedule_state: Option<AssignmentScheduleState>,
    cached_assignment: Option<CachedAssignment>,
}

#[derive(Debug, Clone)]
struct AssignmentScheduleState {
    slots_of_choices: Vec<Option<usize>>,
    scheduled_choices_by_slot: Vec<Vec<usize>>,
    covered_choosers_by_slot: Vec<u32>,
    blocked_edges: HashSet<(usize, usize)>,
}

#[derive(Debug, Clone)]
struct CachedAssignment {
    slots_of_choices: Vec<Option<usize>>,
    assignment: Assignment,
    is_optimal: bool,
}

#[derive(Debug)]
enum LimitSolveResult {
    Feasible(Assignment),
    Infeasible,
    Incomplete,
}

impl<'a> AssignmentSolver<'a> {
    /// Creates a solver for the assignment stage of the prepared problem.
    #[must_use]
    pub fn new(problem: &'a PreparedProblem, options: &'a Options) -> Self {
        Self::new_with_progress(problem, options, ProgressReporter::dummy())
    }

    pub(crate) fn new_with_progress(
        problem: &'a PreparedProblem,
        options: &'a Options,
        progress_reporter: ProgressReporter,
    ) -> Self {
        Self {
            problem,
            options,
            progress_reporter,
            lp_count: 0,
            cached_schedule_state: None,
            cached_assignment: None,
        }
    }

    /// Solves the assignment stage for a fixed scheduling.
    #[must_use]
    pub fn solve(&mut self, scheduling: &Scheduling) -> Option<Assignment> {
        self.solve_until(scheduling, None)
    }

    pub(crate) fn solve_until(
        &mut self,
        scheduling: &Scheduling,
        deadline: Option<SystemTime>,
    ) -> Option<Assignment> {
        let input_data = &self.problem.input_data;
        let levels = input_data.preference_levels.clone();
        let start = levels
            .iter()
            .position(|&level| level >= self.problem.critical_set_analysis.preference_bound())
            .unwrap_or_else(|| levels.len().saturating_sub(1));
        let schedule_state = self.prepare_schedule_state(scheduling);

        if let Some(cached) = &self.cached_assignment
            && cached.slots_of_choices == schedule_state.slots_of_choices
            && cached.is_optimal
        {
            return Some(cached.assignment.clone());
        }

        status::debug(&format!(
            "Starting assignment solve with preference search start index {start} and {} level(s).",
            levels.len()
        ));

        if self.options.greedy {
            status::debug("Greedy mode enabled; solving with maximum preference limit.");
            let result = self.solve_with_limit(&schedule_state, input_data.max_preference, deadline);
            let (assignment, is_optimal) = match result {
                LimitSolveResult::Feasible(assignment) => (Some(assignment), true),
                LimitSolveResult::Infeasible => (None, false),
                LimitSolveResult::Incomplete => (None, false),
            };
            return self.finish_assignment_result(
                assignment,
                schedule_state.slots_of_choices,
                is_optimal,
            );
        }

        let mut low = start;
        let mut high = levels.len().checked_sub(1)?;
        let mut best = None;
        let mut is_optimal = true;

        while low <= high {
            let mid = low + (high - low) / 2;
            let preference_limit = levels[mid];
            status::trace(&format!(
                "Trying assignment solve with preference limit {preference_limit} (range {low}..={high})."
            ));
            match self.solve_with_limit(&schedule_state, preference_limit, deadline) {
                LimitSolveResult::Feasible(assignment) => {
                    status::debug(&format!(
                        "Assignment feasible with preference limit {preference_limit}."
                    ));
                    best = Some(assignment);
                    if mid == start {
                        break;
                    }
                    high = mid.saturating_sub(1);
                }
                LimitSolveResult::Infeasible => {
                    status::trace(&format!(
                        "Assignment infeasible with preference limit {preference_limit}."
                    ));
                    low = mid + 1;
                }
                LimitSolveResult::Incomplete => {
                    status::debug(&format!(
                        "Assignment search could not classify preference limit {preference_limit}; continuing with more relaxed limits."
                    ));
                    is_optimal = false;
                    low = mid + 1;
                }
            }
        }

        self.finish_assignment_result(best, schedule_state.slots_of_choices, is_optimal)
    }

    fn finish_assignment_result(
        &mut self,
        assignment: Option<Assignment>,
        slots_of_choices: Vec<Option<usize>>,
        is_optimal: bool,
    ) -> Option<Assignment> {
        if let Some(assignment) = assignment {
            if is_optimal {
                self.cached_assignment = Some(CachedAssignment {
                    slots_of_choices,
                    assignment: assignment.clone(),
                    is_optimal,
                });
            }
            Some(assignment)
        } else {
            None
        }
    }

    fn build_schedule_state_from_scratch(
        &self,
        scheduling: &Scheduling,
    ) -> AssignmentScheduleState {
        let input_data = &self.problem.input_data;
        let slots_of_choices = scheduling.to_data();
        let mut scheduled_choices_by_slot = vec![Vec::new(); input_data.slots.len()];
        let mut covered_choosers_by_slot = vec![0_u32; input_data.slots.len()];

        for (choice, &slot) in slots_of_choices.iter().enumerate() {
            let Some(slot) = slot else {
                continue;
            };
            scheduled_choices_by_slot[slot].push(choice);
            covered_choosers_by_slot[slot] += input_data.choices[choice].min;
        }

        let blocked_edges =
            self.build_blocked_constraint_edges(&slots_of_choices, &scheduled_choices_by_slot);

        AssignmentScheduleState {
            slots_of_choices,
            scheduled_choices_by_slot,
            covered_choosers_by_slot,
            blocked_edges,
        }
    }

    fn update_schedule_state(
        &self,
        previous: &AssignmentScheduleState,
        scheduling: &Scheduling,
    ) -> AssignmentScheduleState {
        let input_data = &self.problem.input_data;
        let slots_of_choices = scheduling.to_data();
        let mut next = previous.clone();

        for (choice, (&old_slot, &new_slot)) in previous
            .slots_of_choices
            .iter()
            .zip(slots_of_choices.iter())
            .enumerate()
        {
            if old_slot == new_slot {
                continue;
            }

            if let Some(slot) = old_slot {
                if let Some(index) = next.scheduled_choices_by_slot[slot]
                    .iter()
                    .position(|&scheduled_choice| scheduled_choice == choice)
                {
                    next.scheduled_choices_by_slot[slot].swap_remove(index);
                }
                next.covered_choosers_by_slot[slot] -= input_data.choices[choice].min;
            }

            if let Some(slot) = new_slot {
                next.scheduled_choices_by_slot[slot].push(choice);
                next.covered_choosers_by_slot[slot] += input_data.choices[choice].min;
            }
        }

        next.slots_of_choices = slots_of_choices;
        next.blocked_edges = self.build_blocked_constraint_edges(
            &next.slots_of_choices,
            &next.scheduled_choices_by_slot,
        );
        next
    }

    fn prepare_schedule_state(&mut self, scheduling: &Scheduling) -> AssignmentScheduleState {
        let next = if let Some(previous) = &self.cached_schedule_state {
            self.update_schedule_state(previous, scheduling)
        } else {
            self.build_schedule_state_from_scratch(scheduling)
        };
        self.cached_schedule_state = Some(next.clone());
        next
    }

    fn build_blocked_constraint_edges(
        &self,
        slots_of_choices: &[Option<usize>],
        scheduled_choices_by_slot: &[Vec<usize>],
    ) -> HashSet<(usize, usize)> {
        let input_data = &self.problem.input_data;
        let static_data = &self.problem.static_flow_data;
        let mut blocked_edges = HashSet::new();

        for constraint in &static_data.constraints {
            match constraint.kind {
                ConstraintType::ChooserIsInChoice => {
                    let constrained_choice = constraint.other_choice();
                    let Some(slot_index) = slots_of_choices[constrained_choice] else {
                        continue;
                    };
                    let from = static_data.chooser_nodes[constraint.left][slot_index];

                    for &choice in &scheduled_choices_by_slot[slot_index] {
                        if choice == constrained_choice {
                            continue;
                        }
                        blocked_edges.insert((from, static_data.choice_nodes[choice]));
                    }
                }
                ConstraintType::ChooserIsNotInChoice => {
                    let blocked_choice = constraint.other_choice();
                    let to = static_data.choice_nodes[blocked_choice];
                    for slot in 0..input_data.slots.len() {
                        blocked_edges
                            .insert((static_data.chooser_nodes[constraint.left][slot], to));
                    }
                }
                ConstraintType::ChoicesHaveSameChoosers
                | ConstraintType::ChoosersHaveSameChoices => {}
                kind => panic!(
                    "This kind of constraint is not compatible with the min cost flow solver: {kind:?}."
                ),
            }
        }

        blocked_edges
    }

    fn create_edge_groups(
        &self,
        schedule_state: &AssignmentScheduleState,
        flow: &mut MipFlow<u64, u64>,
    ) {
        let input_data = &self.problem.input_data;
        let static_data = &self.problem.static_flow_data;
        for constraint in &input_data.assignment_constraints {
            if constraint.kind != ConstraintType::ChoosersHaveSameChoices {
                continue;
            }

            let other_chooser = constraint.other_chooser();
            for slot in 0..input_data.slots.len() {
                for &choice in &schedule_state.scheduled_choices_by_slot[slot] {
                    let from1 = static_data.chooser_nodes[constraint.left][slot];
                    let to1 = static_data.choice_nodes[choice];
                    let from2 = static_data.chooser_nodes[other_chooser][slot];
                    let to2 = static_data.choice_nodes[choice];
                    flow.create_edge_group_or_block_edges([
                        MipFlowStaticData::edge_id(from1, to1),
                        MipFlowStaticData::edge_id(from2, to2),
                    ]);
                }
            }
        }

        for group in &input_data.dependent_choice_groups {
            if group.len() == 1 {
                continue;
            }

            for chooser in 0..input_data.choosers.len() {
                let mut edge_group = Vec::new();
                for choice in group.iter().copied() {
                    let Some(slot_index) = schedule_state.slots_of_choices[choice] else {
                        continue;
                    };
                    let from = static_data.chooser_nodes[chooser][slot_index];
                    let to = static_data.choice_nodes[choice];
                    edge_group.push(MipFlowStaticData::edge_id(from, to));
                }
                flow.create_edge_group_or_block_edges(edge_group);
            }
        }
    }

    fn solve_with_limit(
        &mut self,
        schedule_state: &AssignmentScheduleState,
        preference_limit: u32,
        deadline: Option<SystemTime>,
    ) -> LimitSolveResult {
        let input_data = &self.problem.input_data;
        let static_data = &self.problem.static_flow_data;
        if deadline.is_some_and(|deadline| deadline <= SystemTime::now()) {
            status::debug("Skipping assignment solve because the worker deadline is exhausted.");
            return LimitSolveResult::Incomplete;
        }

        let mut flow = static_data.base_flow.clone();
        let mut blocked_edges = schedule_state.blocked_edges.clone();
        blocked_edges.extend(static_data.blocked_edges.iter().copied());

        for chooser in 0..input_data.choosers.len() {
            for slot in 0..input_data.slots.len() {
                flow.set_supply(static_data.chooser_nodes[chooser][slot], 1);
            }
        }

        for (choice, &slot) in schedule_state.slots_of_choices.iter().enumerate() {
            if slot.is_none() {
                continue;
            }
            flow.set_supply(
                static_data.choice_nodes[choice],
                -i32::try_from(input_data.choices[choice].min).expect("choice min must fit in i32"),
            );
        }

        for slot in 0..input_data.slots.len() {
            flow.set_supply(
                static_data.slot_nodes[slot],
                -(i32::try_from(input_data.choosers.len()).expect("chooser count must fit in i32")
                    - i32::try_from(schedule_state.covered_choosers_by_slot[slot])
                        .expect("covered chooser count must fit in i32")),
            );
        }

        for chooser in 0..input_data.choosers.len() {
            for (slot, choices) in schedule_state.scheduled_choices_by_slot.iter().enumerate() {
                let from = static_data.chooser_nodes[chooser][slot];
                for &choice in choices {
                    if input_data.choosers[chooser].preferences[choice] > preference_limit {
                        continue;
                    }

                    let to = static_data.choice_nodes[choice];
                    if blocked_edges.contains(&(from, to)) {
                        continue;
                    }
                    let cost = ((f64::from(input_data.choosers[chooser].preferences[choice]) + 1.0)
                        .powf(self.options.preference_exponent))
                        as i64;
                    flow.add_keyed_edge(MipFlowStaticData::edge_id(from, to), from, to, 1, cost);
                }
            }
        }

        for (choice, &slot_index) in schedule_state.slots_of_choices.iter().enumerate() {
            let Some(slot_index) = slot_index else {
                continue;
            };
            flow.add_keyed_edge(
                MipFlowStaticData::edge_id(
                    static_data.choice_nodes[choice],
                    static_data.slot_nodes[slot_index],
                ),
                static_data.choice_nodes[choice],
                static_data.slot_nodes[slot_index],
                i32::try_from(input_data.choices[choice].max - input_data.choices[choice].min)
                    .expect("choice width must fit in i32"),
                0,
            );
        }

        self.create_edge_groups(schedule_state, &mut flow);

        self.lp_count += 1;
        self.progress_reporter.publish_lp(self.lp_count);
        status::trace(&format!(
            "Solving assignment LP/MIP instance #{} with {} node(s) and {} edge(s).",
            self.lp_count,
            flow.node_count(),
            flow.edge_count()
        ));
        let remaining = remaining_duration(deadline);
        if remaining.is_some_and(|remaining| remaining.is_zero()) {
            status::debug("Skipping assignment flow solve because no time remains.");
            return LimitSolveResult::Incomplete;
        }
        if !flow.solve(remaining) {
            if deadline.is_some_and(|deadline| SystemTime::now() >= deadline) {
                status::debug(
                    "Assignment flow solve stopped before producing a feasible solution.",
                );
                return LimitSolveResult::Incomplete;
            }
            status::trace("Assignment flow solve reported infeasibility.");
            return LimitSolveResult::Infeasible;
        }

        let mut data = vec![vec![usize::MAX; input_data.slots.len()]; input_data.choosers.len()];
        for (chooser, slots) in data.iter_mut().enumerate() {
            for (slot, choice_slot) in slots.iter_mut().enumerate() {
                let from = static_data.chooser_nodes[chooser][slot];
                for &choice in &schedule_state.scheduled_choices_by_slot[slot] {
                    let to = static_data.choice_nodes[choice];
                    if flow.solution_value_at(&MipFlowStaticData::edge_id(from, to)) == 1 {
                        *choice_slot = choice;
                        break;
                    }
                }
                assert_ne!(
                    *choice_slot,
                    usize::MAX,
                    "each chooser/slot must be assigned a choice"
                );
            }
        }

        status::trace("Assignment flow solve succeeded.");
        LimitSolveResult::Feasible(Assignment::new(input_data, data))
    }
}

fn remaining_duration(deadline: Option<SystemTime>) -> Option<Duration> {
    deadline.map(|deadline| {
        deadline
            .duration_since(SystemTime::now())
            .unwrap_or(Duration::ZERO)
    })
}
